### DISCUSSION TITLE
Require auth for all pages

### DISCUSSION THREAD

**Discussion by prescottprue** [withastro/starlight#1738] [Feature Requests]
### What version of `starlight` are you using?

0.21.5

### What is your idea?

Ability to apply auth-astro auth check to all pages like so, maybe within a `pages/index.astro` - this may be able to be supported through [this pull request](https://github.com/withastro/starlight/pull/1255):

```astro
---
import type { Props } from '@astrojs/starlight/props';
import Default from '@astrojs/starlight/components/PageFrame.astro';
import { getSession } from 'auth-astro/server';

const session = await getSession(Astro.request)

// If the user is not logged in, redirect them to the login page
if (!session) {
  return Astro.redirect('/login');
}
---

<slot /> // Not sure if this what would be rendered here
```

Its a plus if it could work [as described in the dynamic redirect docs](https://docs.astro.build/en/guides/routing/#dynamic-redirects) like shown above, but this is by no means a requirement. As mentioned, this may be able to be supported through https://github.com/withastro/starlight/pull/1255

## What I've Tried
Both solutions use the following astro config (won't provide `auth.config.mjs` file since it is just config):

```ts
import { defineConfig } from 'astro/config';
import starlight from '@astrojs/starlight';
import auth from 'auth-astro';
import node from '@astrojs/node';

// https://astro.build/config
export default defineConfig({
  output: 'server',
  adapter: node({
    mode: 'standalone',
  }),
  integrations: [
    starlight({
      title: 'Docs',
      // Override PageFrame component only for first solution
      components: {
        PageFrame: './src/PageFrame.astro',
      },
    }),
    auth({}),
  ],
});

```

### Overriding PageFrame Component
I went the route of trying to override PageFrame to check session using auth-astro following [dynamic redirect docs](https://docs.astro.build/en/guides/routing/#dynamic-redirects). I get the error ResponseSentError: The response has already been sent to the browser and cannot be altered. 

Here is my `PageFrame` component:

```astro
---
import type { Props } from '@astrojs/starlight/props';
import Default from '@astrojs/starlight/components/PageFrame.astro';
import { getSession } from 'auth-astro/server';

const session = await getSession(Astro.request)

// If the user is not logged in, redirect them to the login page
if (!session) {
  return Astro.redirect('/login');
}
---

<Default {...Astro.props}><slot /></Default>
```

The error I receive is `The response has already been sent to the browser and cannot be altered.` - this makes sense as this type of logic should live within a page

### Using Session In Component

I'm able to use session within the component to render sign in components inline, but was looking to leverage dynamic redirect so I could have an actual login page (for not having to handle disabling sidebar etc):

```astro
---
import type { Props } from '@astrojs/starlight/props';
import Default from '@astrojs/starlight/components/PageFrame.astro';
import { getSession } from 'auth-astro/server';
import { SignIn, SignOut } from 'auth-astro/components'

const session = await getSession(Astro.request)
---

{
  session
  ? <Default {...Astro.props}><slot /></Default>
  : (
    <Default {...Astro.props}>
      <SignIn provider="okta">
        Sign In
      </SignIn>
      <SignOut>
        Sign Out
      </SignOut>
    </Default>
  )
}
```

### Why is this feature necessary?

This feature would allow anyone trying make their Starlight app be protected by auth (common for corporate documentation and invite only blogs). Since [auth-astro already does a great job of covering dynamic redirect](https://docs.astro.build/en/guides/routing/#dynamic-redirects) auth for basic Astro apps, it would make the feature easier for those familiar with Astro

### Do you have examples of this feature in other projects?

https://docs.astro.build/en/guides/routing/#dynamic-redirects

### Participation

- [ ] I am willing to submit a pull request for this feature.

**Comment by delucis** [comment]
@TheOtterlord has been working on some auth solutions for Starlight recently. Hopefully we’ll have some docs for it soon, but in the meantime there’s this example repo that might be helpful for folks looking into this!

https://github.com/TheOtterlord/starlight-auth

**Comment by HiDeoo** [comment]
Not using `auth-astro`, but sharing [another example](https://github.com/HiDeoo/starlight-better-auth-example) using the Better Auth library to add authentication to a Starlight project.
